File: python/greedy/partition-labels.py
class Solution:
    def partitionLabels(self, s: str) -> List[int]:
        last = defaultdict(int)
        for i in range(len(s)):
            last[s[i]] = i
        finalanswer = []
        addition = ""
        end = last[s[0]]
        for i in range(len(s)):
            if i <= end:
                end = max(end, last[s[i]])
                addition += s[i]
            else:
                finalanswer.append(len(addition))
                addition = s[i]
                end = last[s[i]]

        if addition:
            finalanswer.append(len(addition))

        return finalanswer

File: c++/greedy/partition-labels.cc
class Solution {
public:
    vector<int> partitionLabels(string s) {
        std::unordered_map<char, int> last_index;
        for (int i = 0; i < s.size(); ++i) {
            last_index[s[i]] = i;
        }
        
        std::vector<int> partitions = {};
        int start = 0;
        int end = 0;
        for (int i = 0; i < s.size(); ++i) {
            end = std::max(end, last_index[s[i]]);
            if (i == end) {
                partitions.push_back(end - start + 1);
                start = i + 1;
            }
        }
        return partitions;
    }
};
